### DISCUSSION TITLE
Serialize pattern of bytes without reformatting it?

### DISCUSSION THREAD

**Discussion by moe213** [nlohmann/json#3792] [Q&A]
Hello, how can I serialize this pattern of bytes, so it stays with the same formatting. I tried multiple things already, but it either gets converted into decimals or is adding more backspaces.

`ordered_json Patterns = { };
Patterns["First"]["Pattern"] = "\xE8\x00\x40\xB1\x34\x00\x48\x8B";
Patterns["First"]["Mask"] = "x?xxx?xx";
`

**Comment by nlohmann** [comment]
JSON strings must be valid UTF-8 text, so nlohmann::json can't dump() an arbitrary raw/binary byte sequence as a plain JSON string field without escaping or re-encoding it — that's inherent to the JSON format, not a library limitation. Also note the literal "\xE8\x00\x40\xB1\x34\x00\x48\x8B" contains embedded NUL bytes, so assigning it to a const-char*-backed std::string truncates at the first \x00 before it even reaches the JSON layer. For raw byte patterns like this, use the library's binary type (json::binary_t / json::binary(...)), which round-trips through msgpack/cbor/bson/ubjson without string re-encoding, or base64-encode the bytes yourself for a JSON-text-safe representation. There is no way to make dump() emit raw non-UTF-8 bytes verbatim inside a JSON string, since that would not be valid JSON.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
